google/cloud/firestore_v1/base_batch.py

```diff
@@ -33,14 +33,22 @@ class BaseWriteBatch(object):
     Args:
         client (:class:`~google.cloud.firestore_v1.client.Client`):
             The client that created this batch.
+        write_ctx (bool):
+            Controls whether this instance should call `commit` or `write` upon
+            exiting as a context manager. This parameter has no impact if you
+            do not use the instance as a context manager.
     """
 
-    def __init__(self, client) -> None:
+    def __init__(self, client, write_ctx: bool = False) -> None:
         self._client = client
+        self._write_ctx = write_ctx
```

Comment:
This variable is one of the only interesting parts of this PR. For context, the `[Async]WriteBatch` class can be used as a context manager, and upon exiting such a state, called its most interesting method: `commit`. 

However, this PR adds a sibling method, `write`, which uses a new protobuf object underneath. Thus, how an `[Async]WriteBatch` instance should behave upon exiting its context manager form is no longer clear. The moment of transforming an instance into a context manager does not offer any levers for control, so I opted to add a flag to the constructor.

Open to ideas, but that's the situation :)

Reply:
The other option would be to subclass `BaseWriteBatch` as `BaseBulkWriteBatch` (and the non-base versions), overriding `__exit__` / `__aexit__` to call `self.write()`.  ISTM that might be a clearer signal.  Your split of the testcase class below seems to signal that, too.

Reply:
It wasn't obvious to me whether that was overkill for this one line branch, but maybe it is the right way to go. I'll give it a look.